### PR TITLE
Improve how we avoid fragmentation for large allocations

### PR DESCRIPTION
**Pull request by ronawho**
In #15992 we reduced fragmentation for large allocations by using a dedicated arena to satisfy large allocations. However, the implementation we used was pretty hacky since it was changing the arena for the current thread, allocating, and then changing it back.

This cleans up and improves how we specify which arena to use.  Instead of switching the current thread's arena, just use the extended API which allows us to specify which arena to use in the allocation call.

Note that the jemalloc extended API has undefined behavior for 0 size allocations and NULL frees so so we need to handle those cases ourselves. This adds some API tests to check that the corner cases behave correctly.

This also adds some fragmentation tests and some concurrent allocation stress tests to make sure we're avoiding cross-task fragmentation and not introducing any races.

Lastly, this updates our third-party Makefile to trigger a qthreads rebuild when our memory allocation interface changes. Qthreads uses our allocator so we should have been doing this before, but it was pretty rare for our allocation headers to change so we didn't notice this before.

Closes https://github.com/Cray/chapel-private/issues/1093
Closes https://github.com/cray/chapel-private/issues/1130

### CODE REVIEW

File: runtime/include/mem/jemalloc/chpl-mem-impl.h [hunk 1]
```diff
@@ -35,57 +35,50 @@
...
-  rc = CHPL_JE_POSIX_MEMALIGN(&ret, boundary, size);
-  if( rc == 0 ) return ret;
-  else return NULL;
+  if (size == 0) { size = 1; }
+  return CHPL_JE_MALLOCX(size, MALLOCX_ALIGN(boundary) | CHPL_JE_MALLOCX_ARENA_FLAG(size));
 }
 
 static inline void* chpl_realloc(void* ptr, size_t size) {
-  return CHPL_JE_REALLOC(ptr, size);
+  if (ptr == NULL) { return CHPL_JE_MALLOCX(size, CHPL_JE_MALLOCX_ARENA_FLAG(size)); }
```
Comment (gbtitus) [inline runtime/include/mem/jemalloc/chpl-mem-impl.h:74]: Need a `size==0` check here, don't we?

File: runtime/include/mem/jemalloc/chpl-mem-impl.h [hunk 2]
```diff
@@ -35,57 +35,50 @@
...
+  }
+  return MALLOCX_NO_FLAGS;
 }
-#endif
+
 
 static inline void* chpl_calloc(size_t n, size_t size) {
-  return CHPL_JE_CALLOC(n,size);
+  size_t nSize = n*size;
+  if (nSize == 0) { nSize = 1; }
```
Comment (gbtitus) [inline runtime/include/mem/jemalloc/chpl-mem-impl.h:59]: Why not just return NULL, instead of going to the trouble of allocating an unwanted byte?

File: runtime/include/mem/jemalloc/chpl-mem-impl.h (hunk 1 above)
Comment (ronawho) [reply runtime/include/mem/jemalloc/chpl-mem-impl.h:74]: Oh, shoot -- yes, good catch. Also need to check for NULL pointer on the dalloc on the next line.

File: runtime/include/mem/jemalloc/chpl-mem-impl.h [hunk 3]
```diff
@@ -35,57 +35,50 @@
...
-  return CHPL_JE_REALLOC(ptr, size);
+  if (ptr == NULL) { return CHPL_JE_MALLOCX(size, CHPL_JE_MALLOCX_ARENA_FLAG(size)); }
+  if (size == 0) { CHPL_JE_DALLOCX(ptr, MALLOCX_NO_FLAGS); return NULL; }
+  return CHPL_JE_RALLOCX(ptr, size, CHPL_JE_MALLOCX_ARENA_FLAG(size));
 }
 
 static inline void chpl_free(void* ptr) {
-  CHPL_JE_FREE(ptr);
+  if (ptr == NULL) { return; }
+  CHPL_JE_DALLOCX(ptr, MALLOCX_NO_FLAGS);
```
Comment (gbtitus) [inline runtime/include/mem/jemalloc/chpl-mem-impl.h:81]: `DALLOC()` doesn't need the arena flag even if the pointer is to the "large" arena instead of the thread's default one?  (I guess it must not, if the tests passed ...)

**Comment by gbtitus** [comment]
I confess I'm a bit put off by all the `if (whatever) { stmt; }` additions.  I guess to the extent we have a runtime coding style it's not written down, but I don't think except in some rare circumstances we've ever done one-line conditionals like that.

File: runtime/include/mem/jemalloc/chpl-mem-impl.h (hunk 2 above)
Comment (ronawho) [reply runtime/include/mem/jemalloc/chpl-mem-impl.h:59]: Our memory API interprets a `NULL` return as OOM, so we return a valid pointer. See also https://github.com/chapel-lang/chapel/issues/5273 for some additional detail

Comment (gbtitus) [reply runtime/include/mem/jemalloc/chpl-mem-impl.h:59]: Oh, of course, thanks.

File: runtime/include/mem/jemalloc/chpl-mem-impl.h (hunk 3 above)
Comment (ronawho) [reply runtime/include/mem/jemalloc/chpl-mem-impl.h:81]: Correct, and even more so dallocx will ignore any arena flag passed to it anyways -- https://github.com/jemalloc/jemalloc/blob/e6cb7a1c9b31de3c6eca367d9164a1896bbb60ae/ChangeLog#L940

**Comment by ronawho** [comment]
> I confess I'm a bit put off by all the if (whatever) { stmt; } additions. I guess to the extent we have a runtime coding style it's not written down, but I don't think except in some rare circumstances we've ever done one-line conditionals like that.

I'll change these. Do you have a preference between things like:

```c
if (size == 0) size = 1;
```

```c
if (size == 0) 
  size = 1;
```
```c
if (size == 0) { 
  size = 1;
}
```

**Comment by gbtitus** [comment]
> Do you have a preference ...

I think the last is what we've always done, though for sure it's unofficial.  The single-line style is a greater concern for me than the curlies, even though historically it's the latter that leads to hard-to-spot bugs.

File: runtime/include/mem/jemalloc/chpl-mem-impl.h (hunk 3 above)
Comment (gbtitus) [reply runtime/include/mem/jemalloc/chpl-mem-impl.h:81]: That's a relief!

**Comment by ronawho** [comment]
Fixed the size=0 realloc case, and got rid of the single-line branches. (You can click on the `force-pushed` link to see the diff)

I should have mentioned this originally, but this is probably easiest to review commit-by-commit
